libraries/AP_RangeFinder/AP_RangeFinder_LightWareSerial.cpp

```diff
@@ -38,25 +38,68 @@ bool AP_RangeFinder_LightWareSerial::get_reading(uint16_t &reading_cm)
     int16_t nbytes = uart->available();
     while (nbytes-- > 0) {
         char c = uart->read();
-        if (c == '\r') {
-            linebuf[linebuf_len] = 0;
-            const float dist = strtof(linebuf, NULL);
-            if (!is_negative(dist)) {
-                sum += dist;
-                valid_count++;
-            } else {
-                invalid_count++;
-            }
-            linebuf_len = 0;
-        } else if (isdigit(c) || c == '.' || c == '-') {
-            linebuf[linebuf_len++] = c;
-            if (linebuf_len == sizeof(linebuf)) {
-                // too long, discard the line
+
+        // use legacy protocol
+        if (protocol_state == ProtocolState::UNKNOWN || protocol_state == ProtocolState::LEGACY) {
+            if (c == '\r') {
+                linebuf[linebuf_len] = 0;
+                const float dist = strtof(linebuf, NULL);
+                if (!is_negative(dist)) {
+                    sum += dist;
+                    valid_count++;
+                    // if still determining protocol update legacy valid count
+                    if (protocol_state == ProtocolState::UNKNOWN) {
+                        legacy_valid_count++;
+                    }
+                } else {
+                    invalid_count++;
+                }
                 linebuf_len = 0;
+            } else if (isdigit(c) || c == '.' || c == '-') {
+                linebuf[linebuf_len++] = c;
+                if (linebuf_len == sizeof(linebuf)) {
+                    // too long, discard the line
+                    linebuf_len = 0;
+                }
+            }
+        }
+
+        // use binary protocol
+        if (protocol_state == ProtocolState::UNKNOWN || protocol_state == ProtocolState::BINARY) {
+            bool msb_set = BIT_IS_SET(c, 7);
+            if (msb_set) {
+                // received the high byte
+                high_byte = c;
+                high_byte_received = true;
+            } else {
+                // received the low byte which should be second
+                if (high_byte_received) {
+                    const float dist = (high_byte & 0x7f) << 7 | (c & 0x7f);
```

Comment:
I believe the low byte has a value of 0x0-0xff.

Reply:
@muramura, thanks for the review.  I don't think the low byte could have a value up to 0xff because that would require the top bit to be "1" and then it couldn't be distinguished from the high byte.